core/src/internal/rpc/mod.rs

```diff
@@ -3,25 +3,74 @@ use scaii_defs::protos::{MultiMessage, ScaiiPacket};
 use scaii_defs::protos::{ModuleInit, RpcConfig};
 use std::error::Error;
 use std::net::{IpAddr, SocketAddr};
-
 use websocket::client::sync::Client;
 use websocket::stream::sync::TcpStream;
 
 use scaii_defs::protos::endpoint::Endpoint;
-
+use std::process::Child;
+use std::process::Command;
 use super::LoadedAs;
+use scaii_defs::protos::init_as::InitAs;
+use scaii_defs::protos;
+use std::thread;
+use std::sync::mpsc;
 
 #[cfg(test)]
 mod test;
 
 pub fn init_rpc(rpc_config: RpcConfig) -> Result<LoadedAs, Box<Error>> {
-    use scaii_defs::protos::init_as::InitAs;
-    use scaii_defs::protos::endpoint::Endpoint;
-    use scaii_defs::protos;
+    let (tx, rx) = mpsc::channel();
+    let command = rpc_config.command.clone();
+    match command {
+        None => (),
+        Some(cmd) => {
+            let args = rpc_config.command_args.clone();
+            let _handle = thread::spawn(move || {
+                launch_far_end(&cmd, args.to_vec());
+                rx.recv().unwrap();
+            });
+        }
+    }
 
+    let result = startup_module(&rpc_config);
+    tx.send(String::from("rpc_started")).unwrap();
+    result
+}
+// pub fn init_rpc_broken(rpc_config: RpcConfig) -> Result<LoadedAs, Box<Error>> {
+//     let (tx, rx) = mpsc::channel();
+//     let command = rpc_config.command.clone();
+//     let args = rpc_config.command_args.clone();
+//     let handle = thread::spawn(move || match command {
+//         None => (()),
+//         Some(command_string) => {
+//             let mut child = launch_far_end(&command_string, args.to_vec());
+//             while let Err(_) = rx.try_recv() {
+//                 //let exit_status: Result<Option<ExitStatus>> = child.try_wait();
+//                 let exit_status = child.try_wait();
+//                 match exit_status {
+//                     Ok(Some(_status)) => {
+//                         // process must have stopped - try to restart it
+//                         launch_far_end(&command_string, args.to_vec());
+//                     }
+//                     Ok(None) => (),
+//                     Err(err) => panic!(
+//                         "RPC endpoint exited with error: {} - {}",
+//                         command_string,
+//                         err
+//                     ),
+//                 }
+//             }
+//         }
+//     });
+//     let result = startup_module(&rpc_config);
+//     tx.send(String::from("rpc_started")).unwrap();
+//     handle.join().unwrap();
+//     result
+// }
+fn startup_module(rpc_config: &RpcConfig) -> Result<LoadedAs, Box<Error>> {
     let client = connect(&rpc_config)?;
-
-    match rpc_config.init_as.init_as.ok_or_else::<Box<Error>, _>(|| {
+    let init_as = rpc_config.init_as.clone();
```

Comment:
It's not a huge deal, but why are you cloning it?

Reply:
I think I was led there by an ownership complaint, but will check when repositioning the branches

Reply:
Unless this is mine, whatever

Reply:
It's not a big deal. `init_as` isn't large so the clone is cheap, I just avoid `clone` where possible.